src/metrics/cognitive.rs

```diff
@@ -0,0 +1,808 @@
+use serde::ser::Serializer;
+use serde::Serialize;
+use std::fmt;
+use tree_sitter::Node;
+
+use crate::checker::Checker;
+use crate::*;
+
+// TODO: Find a way to increment the cognitive complexity value
+// for recursive code.
+// The method SHOULD work for EACH of the supported languages.
+// A possible solution could be searching for a crate which implements
+// a language interpreter.
+
+/// The `Cognitive Complexity` metric.
+#[derive(Debug)]
+pub struct Stats {
+    structural: usize,
+    nesting: usize,
+    boolean_seq: BoolSequence,
+}
+
+impl Default for Stats {
+    fn default() -> Self {
+        Self {
+            structural: 0,
+            nesting: 0,
+            boolean_seq: BoolSequence::default(),
+        }
+    }
+}
+
+impl Serialize for Stats {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_f64(self.cognitive())
+    }
+}
+
+impl fmt::Display for Stats {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.cognitive())
+    }
+}
+
+impl Stats {
+    /// Merges a second `Cognitive Complexity` metric into the first one
+    pub fn merge(&mut self, other: &Stats) {
+        self.structural += other.structural;
+    }
+
+    /// Returns the `Cognitive Complexity` metric value
+    pub fn cognitive(&self) -> f64 {
+        self.structural as f64
+    }
+}
+
+#[doc(hidden)]
+pub trait Cognitive
+where
+    Self: Checker,
+{
+    fn compute(_node: &Node, _stats: &mut Stats) {}
+}
+
+macro_rules! compute_booleans {
+    ($node: ident, $stats: ident, $( $typs: pat )|*) => {
+        let mut cursor = $node.walk();
+        for child in $node.children(&mut cursor) {
+            if let $( $typs )|* = child.kind_id().into() {
+                $stats.structural = $stats
+                    .boolean_seq
+                    .eval_based_on_prev(child.kind_id(), $stats.structural);
+            }
+        }
+    };
+}
+
+macro_rules! nesting_levels {
```

Comment:
I wonder if we could use a stack in spaces.rs like the one we've for spaces themselves to avoid to walk too much in the tree.

Reply:
I propose to iterate over this in another PR. What do you think?